### ISSUE TITLE
Introduce infrastructure for consistent typography 

### ISSUE THREAD

**Issue by kuasha420** [google/site-kit-wp#8581]
## Feature Description

As discussed [here](https://github.com/google/site-kit-wp/pull/8492#discussion_r1555019837), we should introduce infrastructure that allows for consistent and efficient usage of typography. There are typography variants defined in [Figma](https://www.figma.com/file/ZL8qlbUyR9lmBaF95t0xpq/sitekit-system-gm2%2B?type=design&mode=design&t=pCrNuKrin29nldsU-0), which is used in the design. We should have some sort of reusable way to use these variants in Site Kit. Besides consistency, this will improve DX and result in less duplicated code.

The actual implementation is up for discussion. We can create `Text` components like it's done [here](https://callstack.github.io/react-native-paper/docs/components/Text/), with `variant` props, or we can introduce typography related classes or CSS mixins. 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A reusable `Typography` component is introduced to standardise text styles across the plugin
* Said component accepts prop for `type`, which follows types defined in the [Figma](https://www.figma.com/design/ZL8qlbUyR9lmBaF95t0xpq/sitekit-system-gm2-?node-id=398-250&t=sHTWZF4qKbp85cNe-0), which are `display`, `headline`, `title`, `label` and `body`
  * Each type should follow the style guidance defined in referenced Figma file
* Said component accepts prop for `size` which follows the styles for `small`, `medium` and `large` sizes in the referenced Figma design.
* Respective story is created to list the available variations in a catalog display, similar to the existing [type scale story](https://storage.googleapis.com/site-kit-storybook/develop/index.html?path=/story/type-scale--default-type-scale)

## Implementation Brief

* [x] Add `assets/js/components/Typography.js`
  *  Add `type` prop, enum type, one of `headline`, `title`, `label` or `body`
  * Add `size` prop, enum type, one of `large`, `medium` or `small`
  * Add `as` prop, that accepts string - for Html tag, it can be enum to whitelist `p`, and `h1`-`h6`, while `span` should be set as default. Wrap the output with it, and  use `googlesitekit-typography` as default. Example in the props `{as: Component = 'span', ...}` and then wrap the output with `<Component classnames={classes}>...</Component>`
  * Include additional classes based on the value for `type` and `size`, eq `googlesitekit-typography--headline googlesitekit-typography--medium`
* [x] Add `assets/sass/components/global/_googlesitekit-typography.scss` 
  * [x] Match the styling of the variations from Figma referenced in the AC 
  * [x] You can extract type scales from `storybook/assets/sass/stories/type-scale.scss` 

### Test Coverage

* Add variations in a catalog display, similar to the existing [type scale story](https://storage.googleapis.com/site-kit-storybook/develop/index.html?path=/story/type-scale--default-type-scale). , no need for the VRT to be added

## QA Brief

* Verify that typography items added in the story (https://storage.googleapis.com/site-kit-storybook/develop/index.html) - search for `Typography`, are matching the referenced Figma styles
  * Note: link type is not included here

## Changelog entry

* Introduce consistent `Typography` component.

**Comment by eugene-manuilov** [comment]
AC ✔

**Comment by techanvil** [comment]
Hi @zutigrm, thanks for drafting the IB which is looking good. A couple of points:

- We shouldn't hardwire the container to be a `div`, we might well want it to be a `span`, `p` etc; I'd suggest it should default to a `span` for inline text but allow the tag to be passed in as a prop which overrides the default (for example, Chakra [uses the `as` prop](https://chakra-ui.com/docs/components/box#as-prop) to specify the tag/component, while Material UI [uses the `component` prop](https://mui.com/material-ui/api/box/#props)).
- We should ensure we start using this new component - either by adding a bit of usage via refactors for existing components in this issue, or creating a new issue to start doing that, in order to encourage its use. Please take one of these approaches.

**Comment by zutigrm** [comment]
Thanks @techanvil 

> We shouldn't hardwire the container to be a div, we might well want it to be a span, p etc; I'd suggest it should default to a span for inline text but allow the tag to be passed in as a prop which overrides the default (for example, Chakra [uses the as prop](https://chakra-ui.com/docs/components/box#as-prop) to specify the tag/component, while Material UI [uses the component prop](https://mui.com/material-ui/api/box/#props)).

Good shout, I updated that part to introduce `as` prop

> We should ensure we start using this new component - either by adding a bit of usage via refactors for existing components in this issue, or creating a new issue to start doing that, in order to encourage its use. Please take one of these approaches.

I did open follow up yesterday, it is marked as blocked by this issue. I will past the reference here as well #10905. Perhaps you can help move it with an AC, since I don't think it fits the criteria of being straightforward one, hence I can't pick it up, but I can pick it up for IB once it lands

**Comment by techanvil** [comment]
Thanks @zutigrm! And, apologies - I missed that #10905 had been added a blocked issue. That's great. I'm a bit busy with SFR design doc work at the moment, so I've moved 10905 to AC, I'll update it myself when time allows but leave it unassigned so someone else can pick it up in the meantime if they have capacity.

IB ✅

**Comment by tofumatt** [comment]
QA is already done as part of Code Review, so moving directly to Approval.
